### PR TITLE
feat(bitops_family): implement bitfield command

### LINKED ISSUE #213: support BITFIELD command
### Discussed in https://github.com/dragonflydb/dragonfly/discussions/212

<div type='discussions-op-text'>

<sup>Originally posted by **koron** July 20, 2022</sup>
Are there any plans to support the [BITMAP commands group](https://redis.io/commands/?group=bitmap)?

</div>

### Summary

I want to use [`BITFIELD` command](https://redis.io/commands/bitfield/) for specifalized bloom filter [koron/java-vbf](https://github.com/koron/java-vbf). It uses `BITFIELD` to get/set partial bits of a string, with combining transaction (`MULTI` / `EXEC`).

### How can you test with java-vbf

1. Install JDK 8 or higher (ex. [OpenJDK 18](https://jdk.java.net/18/))
2. Install [gradle](https://gradle.org/install/)
3. Run redis (or dragonfly) on `localhost:6379`
4. Checkout https://github.com/koron/java-vbf
    ```console
    $ git clone https://github.com/koron/java-vbf.git
    ```
5. Run tests with gradle wrapper
    ```console
    $ ./gradlew test --tests '*.vbf3.*'
    ```
    It will take about 20 seconds.
    It will pass with redis, but fail with dragonfly, because dragonfly doesn't support `BITFIELD` command yet.

### More details

As you can see in https://redis.io/commands/bitfield/, BITFIELD command is a bit complex. But java-vbf uses in two usecases.

1. Get multiple partial (u8) values from a key ([code](https://github.com/koron/java-vbf/blob/b007e91e63df72924f09ffbc085faa24381f19ae/src/main/java/net/kaoriya/vbf3/RedisVBF3.java#L246-L264))
    Example command:
    ```
    BITFIELD somekey GET u8 0 GET u8 32 GET u8 48
    ```
2. Set multiple partial (u8) values for a key ([code](https://github.com/koron/java-vbf/blob/b007e91e63df72924f09ffbc085faa24381f19ae/src/main/java/net/kaoriya/vbf3/RedisVBF3.java#L283-L291))

    Example command:
    ```
    BITFIELD somekey SET u8 0 111 SET u8 32 222 SET u8 48 333
    ```

Please keep in mind that java-vbf does not require all of BITFILED's functionality to run.

### PR DESCRIPTION
**Pull request by kostasrim**
Resolves #213 

* implement bitfield command
* add tests

Note that there are certain optimizations that can be done. 

1. We read/write bit by bit (this is done in redis as well) even when we address the whole byte. Therefore, to update a single byte, we need 8 ops. To update 8 bytes, we need 64 ops. We can do that in one op, when these are aligned. Unless this becomes a proven bottleneck I would not spent time optimizing it further.

2. I do not like that we copy the whole string (target key value -- i.e, the bitfield) when we start processing/executing the command. This is wasteful, and for large large bitfields, this can quickly become a bottleneck. However, doing otherwise requires a little bit more involved work. In particular, we need a way to fetch and store individual bytes from the `CompactObject` and the interfaces `GetString()` and `SetString()` are not sufficient. Again, when and if this ever becomes a bottleneck, we can patch this. I am just mentioning this here for completeness.

P.s. apologies for the large PR. I did my best to keep it as short as possible.

### CODE REVIEW

File: src/server/bitops_family.cc [hunk 1]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+enum class EncodingType { UINT, INT, NILL };
+
+struct CommonAttributes {
+  EncodingType type;
+  size_t encoding_bit_size;
+  size_t offset;
+};
+
+struct Nill {};
+using ResultType = std::variant<int64_t, Nill>;
```
Comment (romange) [inline src/server/bitops_family.cc:605]: why is it needed?

File: src/server/bitops_family.cc [hunk 2]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+          return false;
+      }
+      return true;
+    }
+
+    *value = incr_value + *value;
+    return true;
+  }
+
+  bool IntOverflow(int64_t* value, size_t total_bits, int64_t incr = 0, bool add = false) const {
```
Comment (romange) [inline src/server/bitops_family.cc:648]: I personally prefer that long functions will be defined outside of the class definition to improve readability.
Also, output parameter.

File: src/server/bitops_family.cc [hunk 3]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+ private:
+  CommonAttributes attr_;
+};
+
+class Set {
+ public:
+  explicit Set(CommonAttributes attr, int64_t value) : attr_(attr), set_value_(value) {
+  }
+
+  ResultType ApplyTo(std::string& bytes, Overflow ov) {
```
Comment (romange) [inline src/server/bitops_family.cc:741]: we pass mutable arguments by pointer everywhere in the code.

Comment (romange) [reply src/server/bitops_family.cc:741]: they should be last ones as well

File: src/server/bitops_family.cc [hunk 4]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+
+  Policy type = WRAP;
+};
+
+class Get {
+ public:
+  explicit Get(CommonAttributes attr) : attr_(attr) {
+  }
+
+  ResultType ApplyTo(const std::string& value, Overflow ov) {
```
Comment (romange) [inline src/server/bitops_family.cc:704]: here and below as well: please add comments

**Review by romange — COMMENTED** [review]
Hi Kostas, I did not go over the correctness of the code, just provided some style comments

File: src/server/bitops_family.cc [hunk 5]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+    using namespace std::string_view_literals;
+    if (op == "OVERFLOW"sv) {
+      auto policy = parser.ToUpper().Next();
+      if (auto res = Overflow::PolicyFromStr(policy); res) {
+        result.push_back(Overflow{*res});
+        continue;
+      }
+      return make_unexpected(kSyntaxErr);
+    }
+
```
Comment (dranikpg) [inline src/server/bitops_family.cc:995]: 1. We have `parser.Check()` to simplify tag checking
2. We have `parser.Switch()` so all of the ` Overflow::PolicyFromStr` is not needed

File: src/server/bitops_family.cc [hunk 6]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+  auto value = parser.ToUpper().Next();
+  int64_t res{0};
+  if (!absl::SimpleAtoi(value, &res)) {
+    return make_unexpected(kSyntaxErr);
+  }
+  return res;
+}
+
+nonstd::expected<CommandList, std::string> ParseToCommandList(CmdArgList args) {
```
Comment (dranikpg) [inline src/server/bitops_family.cc:973]: The main idea of silent failures in the parser was to have simple and concise flow without constant checks, so we can just check at the very end if any errors occurred

File: src/server/bitops_family.cc [hunk 7]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+    return make_unexpected(kSyntaxErr);
+  }
+
+  auto value = parser.ToUpper().Next();
+  int64_t res{0};
+  if (!absl::SimpleAtoi(value, &res)) {
+    return make_unexpected(kSyntaxErr);
+  }
+  return res;
+}
```
Comment (dranikpg) [inline src/server/bitops_family.cc:971]: This is just `parser.Next<int64_t>()`... Please see my previous comment about silent failures, because without it might be inconvenient to actually use the parser

File: src/server/bitops_family.cc [hunk 8]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+      return WRAP;
+    }
+    if (policy == "SAT") {
+      return SAT;
+    }
+    if (policy == "FAIL") {
+      return FAIL;
+    }
+    return {};
+  }
```
Comment (dranikpg) [inline src/server/bitops_family.cc:621]: Same here, `parser.Switch()`... (silent failure issue)

File: src/server/bitops_family.cc [hunk 9]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+
+  if (parsed.encoding_bit_size == 64 && parsed.type == EncodingType::UINT) {
+    return make_unexpected(
+        "invalid bitfield type. use something like i16 u8. note that u64 is not supported but i64 "
+        "is.");
+  }
+
+  auto offset_proxy = parser.Next();
+  std::string_view offset_str = offset_proxy;
```
Comment (dranikpg) [inline src/server/bitops_family.cc:944]: `string_view offset_str = parser.Next<string_view>()`

File: src/server/bitops_family.cc [hunk 10]
```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+          int64_t res = add ? incr + 1 : *value % total_elements;
+          *value = max + res;
+          break;
+        }
+        case Overflow::SAT:
+          *value = min;
+          break;
+        case Overflow::FAIL:
+          *value = 0;
+          return false;
```
Comment (dranikpg) [inline src/server/bitops_family.cc:685]: Same code as above but with +1 instead of -1... Maybe we can move this to a function safely

File: src/server/bitops_family.cc (hunk 1 above)
Comment (dranikpg) [reply src/server/bitops_family.cc:605]: to use visit instead of if/else. But in many cases it seems like if/else is simpler 😅

**Comment by dranikpg** [comment]
Actually really sad that C++ has no standard way to work with overflow (Rust has functions for all of it 🤓)

I only know of compiler extensions like this https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

File: src/server/bitops_family.cc (hunk 1 above)
Comment (romange) [reply src/server/bitops_family.cc:605]: I prefer using `if/else` than using `visit` @kostasrim

**Comment by romange** [comment]
> Actually really sad that C++ has no standard way to work with overflow (Rust has functions for all of it 🤓)
> 
> I only know of compiler extensions like this https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
> 
> Actually it works with clang 👀 https://clang.llvm.org/docs/LanguageExtensions.html#checked-arithmetic-builtins

if gcc/clang extensions can help here to reduce amount of code - I am prefer using them. We are not going to use any other compilers any time soon.

**Comment by dranikpg** [comment]
the main benefit is that compiler overflow/carry functions are guaranteed to be correct

File: src/server/bitops_family.cc (hunk 1 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:605]: We need the special value `Nill`, to denote that one of the subcommands (GET|SET|INCRBY) failed because of overflow/underflow. In that case, we still need to return the value of an empty array. So for example:

```BITFIELD foo SET i1 0 1 OVERFLOW FAIL INCRBY i1 0 1``` 
Will return `{0, null}`. We can't use an `int` or `0` to represent overflow failures and therefore I introduced this Nill value for commands that failed the check. 

I can use an `if/else`, it won't make much of a difference IMO.

```
  1015     auto nill_case = [&](Nill nill) { (*cntx)->SendNull(); };                                                                   
  1016     auto int_cases = [&](auto num) { (*cntx)->SendLong(num); };                                                                 
  1017                                                                                                                                 
  1018     std::visit(Overloaded{nill_case, int_cases}, elem);  
```

vs

```
if(std::holds_alternative<Nill>(elem)) {
    (*cntx)->SendNull();
}
(*cntx)->SendLong(num);
```

But yeah, I will make the change since you find it more helpful

File: src/server/bitops_family.cc (hunk 2 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:648]: done

File: src/server/bitops_family.cc (hunk 5 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:995]: +1 for `parser.check()`. However, the reason I did not user `Switch` was a) I find it a little bit more verbose b) I did not want to check if the parser failed (I find the optional a bit more expressive in this case).

File: src/server/bitops_family.cc (hunk 6 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:973]: > The main idea of silent failures in the parser was to have simple and concise flow without constant checks

Yes I understand that. The reason I used the parser, was for convenience over iterating the arguments (because of Next(), Check() etc) and not because of the silent failures. In some places, it makes sense, and I use that (including your suggestion) but in some other I really do need to manually do the error handling. It's a hybrid approach, but far more better than dealing with the arguments and indexes myself.

File: src/server/bitops_family.cc (hunk 7 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:971]: Yes, this works here nicely, in fact, `ParseValue` can be eliminated all together(which I did).

File: src/server/bitops_family.cc (hunk 8 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:621]: As I said above, I did not use this partly because of the verbosity but also because the silent failure is not very helpful here.

File: src/server/bitops_family.cc (hunk 9 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:944]: nice!

File: src/server/bitops_family.cc (hunk 10 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:685]: done

**Comment by kostasrim** [comment]
> Actually really sad that C++ has no standard way to work with overflow (Rust has functions for all of it 🤓)
> 
> I only know of compiler extensions like this https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html
> 
> Actually it works with clang 👀 https://clang.llvm.org/docs/LanguageExtensions.html#checked-arithmetic-builtins

@romange @dranikpg 

I am aware of those functions and I agree it's sad that C++ doesn't offer builtins. The problems with the compiler extensions you just linked, is that from my understanding I quote:


>These built-in functions promote the first two operands into infinite precision signed type and perform addition on those promoted operands. The result is then cast to the type the third pointer argument points to and stored there. If the stored result is equal to the infinite precision result, the built-in functions return false, otherwise they return true. As the addition is performed in infinite signed precision, these built-in functions have fully defined behavior for all argument values.

This won't do what we like for certain cases. For example, it could be the case that we do:

`BITWISE foo SET i5 5 15`, how are these functions will help with overflowing the `5bit` integer ? How would you then cast the result to a 5 bit number? Would it even work? It has to be done manually unless I am missing something :(

File: src/server/bitops_family.cc (hunk 5 above)
Comment (dranikpg) [reply src/server/bitops_family.cc:995]: I'd suggest then to check at the end that the parser error is null (it fails silently so to check your wrapper)

File: src/facade/cmd_arg_parser.h [hunk 11]
```diff
@@ -143,6 +143,13 @@ struct CmdArgParser {
     return std::exchange(error_, {});
   }
 
+  bool HasAtLeast(size_t i) const {
+    if (i == 0) {
```
Comment (BorysTheDev) [inline src/facade/cmd_arg_parser.h:147]: it looks strange for me, we always should have 0 elements

File: src/server/bitops_family.cc (hunk 8 above)
Comment (BorysTheDev) [reply src/server/bitops_family.cc:621]: you can add methods with explicit failure

**Comment by dranikpg** [comment]
Ah, yes, I see that bitfield allows exotic types like i33 😑 Then I agree that using compiler function is not that productive

One option is always using extension functions with (u)int64 and then clamp'ing the value to 1<<width if needed. At least this makes sure the code is safe

File: src/server/bitops_family.cc (hunk 8 above)
Comment (dranikpg) [reply src/server/bitops_family.cc:621]: How is it more verbose if its 2 lines vs 10 😆But your call after all

Comment (kostasrim) [reply src/server/bitops_family.cc:621]: welll that's also true :rofl:

File: src/facade/cmd_arg_parser.h (hunk 11 above)
Comment (kostasrim) [reply src/facade/cmd_arg_parser.h:147]: +1

**Comment by kostasrim** [comment]
> One option is always using extension functions with (u)int64 and then clamp'ing the value to 1<<width if needed. At least this makes sure the code is safe

I kinda do this, without the extension functions. The thing is: 

> Otherwise, the builtin will return 1 and the result will be equal to the unique value that is equivalent to the mathematically-correct result modulo two raised to the k power, where k is the number of bits in the result type

We still need to handle the cases a little differently. So for a 4 bit type i4, we would need (a + b) mod 2^k, with k = 64. However, what we really need is (a + b) mod 2 ^ k, with k = 4. If (a+b) = 300, then 300 mod 2 ^ 4 = 12. For k = 64, we would get 300 mod 2^64 == 300.

File: src/server/bitops_family.cc (hunk 1 above)
Comment (kostasrim) [reply src/server/bitops_family.cc:605]: The differences are:

1. std::variant is more direct than optional. I either have a `Nill` or a `int64_t`.
2. We can extend in the future by adding overloads without needing to add extra if's. (for now we are feature complete, but we could have new subcommands in the future like MULT, etc which would be far better expressed by a variant)

`Nill` is not `std::monostate` even if they look alike. The first represents a result of a computation. The later is used to allow a std::variant default constructable when the underline types are not default constructor and it HAS to be the first argument in the type list. For example:
```
struct non_default {
   non_default(int) {}
};

std::variant<non_default> var; //compile error
std::variant<non_default, std::monostate> var; //compile error, the default constructor always used the first type argument
std::variant<std::monostate, non_default> var; //Good compiles
```
long story short, I don't want to hijack monostate :stuck_out_tongue: 

But, I am happy to substitute to an if/else and if in the future we need to add more, we will make the simple change :)

Comment (dranikpg) [reply src/server/bitops_family.cc:605]: std::monostate is _literally_ just an empty struct, they're are the same, just different names

> std::variant is more direct than optional. I either have a Nill or a int64_t
  
What's the justification for optional then 🙂 You either carry information or you don't...

Either way, just answering

Comment (kostasrim) [reply src/server/bitops_family.cc]: > std::monostate is literally just an empty struct, they're are the same, just different names

Yes, all I am trying to say is that I was trying to be explicit about the intent. `std::monostate` does not mean `Nill` if you don't know the code. The way I think about it is that if I have a monostate it means that I just want to default construct a variant. It doesn't have any other meaning. On the other hand when I use Nill, it does have a meaning, it's a `Nill` result -- allowing the variant to default construct is a nice added property but not the point of Nill since once I use monostate + one type it's basically an optional :stuck_out_tongue: 

I am not arguing here, I am trying to explain how my thinking around it.

p.s. I removed the variant, it's now an optional :)

File: src/server/bitops_family.cc [hunk 12]
```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+      }
+      case Overflow::SAT:
+        *value = sat_case;
+        break;
+      case Overflow::FAIL:
+        *value = 0;
+        return false;
+    }
+    return true;
+  };
```
Comment (dranikpg) [inline src/server/bitops_family.cc:678]: Looks really fragile but (almost?) correct to me 🙂 

For example, are you aware that you implicitly cast `*value` to uint64_t? What happens in case of an underflow when *value is negative?

A few DCHECKS wouldn't hurt

File: src/server/bitops_family.cc [hunk 13]
```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+
+    auto attr = maybe_attr.value();
+    if (op == "GET"sv) {
+      result.push_back(Command(Get(attr)));
+      continue;
+    }
+
+    auto value = parser.ToUpper().Next<int64_t>();
+    if (parser.HasError()) {
+      parser.Error();
```
Comment (dranikpg) [inline src/server/bitops_family.cc:1027]: That is a really sneaky way to clear errors 😄

**Review by dranikpg — APPROVED** [review]

File: src/server/bitops_family.cc [hunk 14]
```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+  if (!HandleOverflow(ov, *res)) {
+    return {};
+  }
+
+  Set set(attr_, *res);
+  set.ApplyTo(ov, &bytes);
+  return *res;
+}
+
+bool IncrBy::HandleOverflow(Overflow ov, int64_t& previous) {
```
Comment (romange) [inline src/server/bitops_family.cc:833]: please use pointers, not references.

File: src/server/bitops_family.cc [hunk 15]
```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+
+using Result = std::optional<ResultType>;
+
+// Visitor for all the subcommand variants. Calls ApplyTo, to execute the subcommand
+class CommandApplyVisitor {
+ public:
+  explicit CommandApplyVisitor(std::string bitfield) : bitfield_(std::move(bitfield)) {
+  }
+
+  Result operator()(Get& get) {
```
Comment (romange) [inline src/server/bitops_family.cc:856]: same here for all the methods

File: src/server/bitops_family.cc [hunk 16]
```diff
@@ -577,14 +592,513 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
...
+
+  if (visitor.ShouldCommit()) {
+    access_.Find(shard_);
+    access_.Commit(visitor.Bitfield());
+  }
+
+  return results;
+}
+
+nonstd::expected<CommonAttributes, std::string> ParseCommonAttr(CmdArgParser& parser) {
```
Comment (romange) [inline src/server/bitops_family.cc:936]: and here

**Review by dranikpg — APPROVED** [review]
I'm surprised the redis code is that complicated as well...

**Review by dranikpg — APPROVED** [review]
I see the auto merge, not sure if others have smth to add... 🙃
